processor/src/parallel/tests.rs

```diff
@@ -0,0 +1,609 @@
+use alloc::{string::String, sync::Arc};
+use std::string::ToString;
+
+use miden_core::{Kernel, Operation, Program, mast::MastForest};
+use miden_utils_testing::get_column_name;
+use pretty_assertions::assert_eq;
+use rstest::{fixture, rstest};
+
+use super::*;
+use crate::{DefaultHost, HostLibrary, fast::FastProcessor};
+
+const DEFAULT_STACK: &[Felt] = &[Felt::new(1), Felt::new(2), Felt::new(3)];
+
+/// The procedure that DYN and DYNCALL will call in the tests below. Its digest needs to be put on
+/// the stack before the call.
+const DYN_TARGET_PROC_HASH: &[Felt] = &[
+    Felt::new(10995436151082118190),
+    Felt::new(776663942277617877),
+    Felt::new(3177713792132750309),
+    Felt::new(10407898805173442467),
+];
+
+/// The digest of a procedure available to be called via an EXTERNAL node.
+const EXTERNAL_LIB_PROC_DIGEST: Word = Word::new([
+    Felt::new(9552974201798903089),
+    Felt::new(993192251238261044),
+    Felt::new(1885027269046469428),
+    Felt::new(8558115384207742312),
+]);
+
+/// This test verifies that the trace generated when executing a program in multiple fragments (for
+/// all possible fragment boundaries) is identical to the trace generated when executing the same
+/// program in a single fragment. This ensures that the logic for generating trace rows at fragment
+/// boundaries is correct, given that we test elsewhere the correctness of the trace generated in a
+/// single fragment.
+#[rstest]
+// Case 1: Tests the trace fragment generation for when a fragment starts in the start phase of a
+// Join node (i.e. clk 4). Execution:
+//  0: JOIN
+//  1:   BLOCK MUL END
+//  4:   JOIN
+//  5:     BLOCK ADD END
+//  8:     BLOCK SWAP END
+// 11:   END
+// 12: END
+// 13: HALT
+#[case(join_program(), 4, DEFAULT_STACK)]
+// Case 2: Tests the trace fragment generation for when a fragment starts in the finish phase of a
+// Join node. Same execution as previous case, but we want the 2nd fragment to start at clk=11,
+// which is the END of the inner Join node.
+#[case(join_program(), 11, DEFAULT_STACK)]
+// Case 3: Tests the trace fragment generation for when a fragment starts in the start phase of a
+// Split node (i.e. clk 5). Execution:
+//  0: JOIN
+//  1:   BLOCK SWAP SWAP END
+//  5:   SPLIT
+//  6:     BLOCK ADD END
+//  9:   END
+// 10: END
+// 11: HALT
+#[case(split_program(), 5, &[ONE])]
+// Case 4: Similar to previous case, but we take the other branch of the Split node.
+//  0: JOIN
+//  1:   BLOCK SWAP SWAP END
+//  5:   SPLIT
+//  6:     BLOCK SWAP END
+//  9:   END
```

Comment:
same here

Reply:
Same idea, but "false" branch is executed